core/trino-spi/src/main/java/io/trino/spi/exchange/ExchangeSourceOutputSelector.java

```diff
@@ -157,6 +160,52 @@ public ExchangeSourceOutputSelector merge(ExchangeSourceOutputSelector other)
                 this.finalSelector && other.finalSelector);
     }
 
+    @Override
+    public String toString()
+    {
+        return new StringJoiner(", ", ExchangeSourceOutputSelector.class.getSimpleName() + "[", "]")
+                .add("version=" + version)
+                .add("values=" + values.entrySet().stream()
+                        .map(entry -> new AbstractMap.SimpleEntry<ExchangeId, String>(entry.getKey(), valuesSliceToString(entry.getValue())))
+                        .collect(toUnmodifiableMap(
+                                AbstractMap.SimpleEntry::getKey,
+                                AbstractMap.SimpleEntry::getValue)))
+                .add("finalSelector=" + finalSelector)
+                .toString();
+    }
+
+    private String valuesSliceToString(Slice values)
```

Comment:
This is decoding the struct stored in a Slice
where is the struct format documented?

Reply:
probably nowwhere :)

Reply:
😱